pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -130,3 +133,51 @@ func toClientControlPlane(cp *tinkerbell.ControlPlane) *clusters.ControlPlane {
 		Other:                       other,
 	}
 }
+
+// ValidateHardware performs a set of validations on the tinkerbell hardware read from the cluster.
+func (r *Reconciler) ValidateHardware(ctx context.Context, log logr.Logger, clusterSpec *c.Spec) (controller.Result, error) {
+	log = log.WithValues("phase", "validateHardware")
+
+	capiCluster, err := controller.GetCAPICluster(ctx, r.client, clusterSpec.Cluster)
+	if err != nil {
+		return controller.Result{}, errors.Wrap(err, "validating tinkerbell hardware")
+	}
+	if capiCluster != nil {
+		// If CAPI cluster exists, the hardware has been validated
+		// and it's possibly already in use so no need to validate it again.
+		log.V(3).Info("CAPI cluster already exists, skipping hardware validations")
+		return controller.Result{}, nil
+	}
+
+	// We need a new reader each time so that the catalogue gets recreated.
+	etcdReader := hardware.NewETCDReader(r.client)
+	if err := etcdReader.NewCatalogueFromETCD(ctx); err != nil {
```

Comment:
This function is just populating the catalog right? It's not doing any validations itself?

Reply:
oh not this is just for populating. It basically reads the objects from etcd and adds it to the catalogue.